WirescaleApi/WireguardManager.cs

```diff
@@ -0,0 +1,55 @@
+﻿using NetTools;
+
+namespace WirescaleApi;
+
+public class WireguardManager
+{
+    private readonly IWgrestApiClient _wgrestApiClient;
+    private readonly IpAddressAllocator _ipAddressAllocator;
+
+    public WireguardManager(IWgrestApiClient wgrestApiClient, IpAddressAllocator ipAddressAllocator)
+    {
+        _wgrestApiClient = wgrestApiClient;
+        _ipAddressAllocator = ipAddressAllocator;
+    }
+
+    public async Task<WireguardPeerRegistrationResult> RegisterNewPeer(string clientPublicKey)
+    {
+        var deviceName = "wg-wirescale";
+        var wireguardDevices = await _wgrestApiClient.GetWgrestDevices();
+        var wirescaleDevice = wireguardDevices.Single(x => x.Name == deviceName);
+
+        var wireguardPeers = await _wgrestApiClient.GetWgrestPeers(wirescaleDevice.Name);
+        if (wireguardPeers.Any(x => x.PublicKey == clientPublicKey))
+        {
+            throw new InvalidOperationException("Public key is already registered as a peer");
+        }
+
+        // We currently only support one (IPv4) network. Throw an exception if device has more or less than that.
+        var wireguardNetworkCidr = wirescaleDevice.Networks.Single();
+
+        // Get the currently registered peer IP ranges
+        var currentPeerIpRanges = wireguardPeers.Select(x => IPAddressRange.Parse(x.AllowedIps.Single()));
+
+        var clientPeerIpAddress = _ipAddressAllocator.AllocateIpAddress(wireguardNetworkCidr, currentPeerIpRanges);
+
+        var wgrestPeer = await _wgrestApiClient.CreateWgrestPeer(wirescaleDevice.Name, new WgrestPeer
```

Comment:
If this, or any of the other wgrest API calls, fails for whatever reason, what will the UX be for the user?

Reply:
This particular wgrest api call is unlikely to fail -- unfortunately, in my opinion, as it probably *should* return an error rather than accepting bad input. If for instance a public key has already been registered and allocated an allowed-ip range, the wgrest API will simply add the `allowed_ips` to the list and return a successful response. I'd probably prefer if it returned an error in that scenario, but either way, this is why I check for an [existing public key and throw an InvalidOperationException if the peer is already registered](https://github.com/runesoerensen/wirescale/blob/b33ff30c4d9cac642457b628a4ba728f81735409/WirescaleApi/WireguardManager.cs#LL23C52-L23C67).

This exception type is actually [handled in the controller](https://github.com/runesoerensen/wirescale/blob/implement-wireguard-peer-registration/WirescaleApi/Controllers/WireguardPeerController.cs#L29) and from the perspective of an API the UX would be an HTTP 400 Bad Request (with the error message "Public key is already registered as a peer")).

Reply:
For other wgrest failing API operations the UX isn't great -- generally speaking they're not handled particularly graceful, but they all will throw exceptions (and in turn, error responses) to the Wirescale API consumer [as they all verify success responses](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpresponsemessage.ensuresuccessstatuscode?view=net-7.0).

So basically we could just handle the [`HttpRequestException`](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httprequestexception?view=net-7.0) in the controller as well, and return a more appropriate response there.

Reply:
There's definitely work to be done on the error/exception/failure handling in several parts of this system. Most of it is very straightforward, but I regret I didn't also handle the duplicate public key error condition with a more specific exception than `InvalidOperationException`. 

As a rule of thumb, the code I write will often throw exceptions/errors, and they'll do so intentionally and loudly until they're handled (rather than silently swallowing error conditions) ;)

Reply:
The last 5 commits (from https://github.com/runesoerensen/wirescale/pull/4/commits/b5237d040d529d7c67788c547b8c7ad6509206d0 and forward) address this -- an HTTP 400 is returned on bad user input/something the API consumer can address. Other exceptions are logged, and an HTTP 500 is returned for those instead.